### PR TITLE
client: ignore the MAY_READ access mode check if __FMODE_EXEC is set

### PR DESCRIPTION
**Pull request by lxbsz**
The execlp(),etc system call will ignore the 'r' permission in VFS,
and allowed a file without 'r' permission still could be excuted if
'x' is set.

The kernel VFS and FUSE will pass the 'O_READONLY | __FMODE_EXEC' to
libcephfs in the open_flags when opening the file and if the
'fuse_default_permissions' option is disabled we can check the
'__FMODE_EXEC' flags. If '__FMODE_EXEC' is set then we should ignore
the 'r' permissio check.

Fixes: https://tracker.ceph.com/issues/61501
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/include/ceph_fs.h [hunk 1]
```diff
@@ -457,13 +457,16 @@ extern const char *ceph_mds_op_name(int op);
...
 #define CEPH_O_RDWR            00000002
+#define CEPH_O_FMODE_EXEC      00000040
 #define CEPH_O_CREAT           00000100
 #define CEPH_O_EXCL            00000200
 #define CEPH_O_TRUNC           00001000
 #define CEPH_O_LAZY            00020000
 #define CEPH_O_DIRECTORY       00200000
 #define CEPH_O_NOFOLLOW        00400000
 
+#define O_FMODE_EXEC	       00000040
```
Comment (dparmar18) [inline src/include/ceph_fs.h:468]:
```suggestion
#define O_FMODE_EXEC	       CEPH_O_FMODE_EXEC
```

Value of `FMODE_EXEC` whether for kernel or libcephfs would be same to maintain uniformity right? Should doing this make sense?

File: src/client/Client.cc [hunk 2]
```diff
@@ -10434,6 +10439,7 @@ int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
 {
   ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
 
+  ldout(cct, 3) << "_read(" << f << ", " <<  offset << dendl;
```
Comment (dparmar18) [inline src/client/Client.cc:10442]:
```suggestion
  ldout(cct, 3) << __func__ << f << ", " <<  offset << dendl;
```

Comment (lxbsz) [reply src/client/Client.cc:10442]: Will fix it.

File: src/include/ceph_fs.h (hunk 1 above)
Comment (lxbsz) [reply src/include/ceph_fs.h:468]: yeah, I just copied this from Linux Kernel VFS. And the VFS will pass this flag to lower level layer, which is the FUSE and then FUSE will pass it to the userspace.

File: src/common/options/mds-client.yaml.in [hunk 3]
```diff
@@ -355,7 +355,7 @@ options:
     permissions checking, instead of relying on the permissions enforcement in
     FUSE. Set to ``false`` together with the ``client acl type=posix_acl``
     option to enable POSIX ACL.
-  default: false
+  default: true
```
Comment (gregsfortytwo) [inline src/common/options/mds-client.yaml.in:358]: We need to keep testing our own permissions checking, and relying on the VFS for permissions checking has led to issues with coherency and liveness in the past, so we really can't turn this option off.

**Review by gregsfortytwo — CHANGES_REQUESTED** [review]

File: src/common/options/mds-client.yaml.in (hunk 3 above)
Comment (lxbsz) [reply src/common/options/mds-client.yaml.in:358]: Yeah, makes sense to me. We really have found several issues with this. I will remove this change since the second commit will fix this issue too.

File: src/client/Client.cc [hunk 4]
```diff
@@ -6019,7 +6023,8 @@ int Client::may_delete(Inode *dir, const char *name, const UserPerm& perms)
   return r;
 }
 
-int Client::may_delete(const char *relpath, const UserPerm& perms) {
+int Client::may_delete(const char *relpath, const UserPerm& perms)
+{
```
Comment (dparmar18) [inline src/client/Client.cc:6027]: this change looks unrelated, i hope this is fine?

Comment (lxbsz) [reply src/client/Client.cc:6027]: It's okay. I just fixed the code style to make it the same with all the other we did.

**Review by dparmar18 — APPROVED** [review]
LGTM

**Comment by dparmar18** [comment]
jenkins test api
